File: src/main/java/bernard/tatin/Threads/Mutex.java
package bernard.tatin.Threads;

import java.util.concurrent.Semaphore;

public class Mutex {
    private final Semaphore mutex = new Semaphore(1, true);

    public Mutex() {

    }

    public void lock() throws InterruptedException {
        mutex.acquire();
    }

    public void unlock() {
        mutex.release();
    }
}

File: kotlin-version/src/main/java/bernard/tatin/Threads/Mutex.kt
package bernard.tatin.Threads

import java.util.concurrent.Semaphore

class Mutex {
    private val mutex = Semaphore(1, true)

    @Throws(InterruptedException::class)
    fun lock() {
        mutex.acquire()
    }

    fun unlock() {
        mutex.release()
    }
}
